File: python/84_Largest_Rectangle_In_Histogram.py
class Solution:
  def largestRectangleArea(self, heights: list[int]) -> int:
    myStack = []
    maxA = 0
    for i, h in enumerate(heights):
      start = i
      while myStack and myStack[-1][1] > h:
        index, height = myStack.pop()
        maxA = max(maxA, height * (i - index))
        start = index
      myStack.append((start, h))
    
    for i, h in myStack:
      maxA = max(maxA, h * (len(heights) - i))
    return maxA

File: c++/84_Largest_Rectangle_In_Histogram.cpp
#include <stack>
#include <utility>
#include <vector>

using namespace std;

class Largest_Rectangle_In_Histogram {
public:
  int largestRectangleArea(vector<int>& heights) {
    stack<pair<int,int>> myStack;
    int maxArea = 0;

    for (int i = 0; i < heights.size(); i++) {
      int start = i;
      while (myStack.size() && myStack.top().second > heights[i]) {
        int index = myStack.top().first;
        int height = myStack.top().second;
        myStack.pop();
        int area = height * (i - index);
        maxArea = max(maxArea, area);
        start = index;
      }
      myStack.push({start,heights[i]});
    }

    while (myStack.size()) {
      int index = myStack.top().first;
      int height = myStack.top().second;
      int area = height * (heights.size() - index);
      maxArea = max(maxArea, area);
      myStack.pop();
    }
    return maxArea;
  }
};